File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(timer) {
  if (timer === undefined) {
    return 'You forgot to set the timer.';
  }
  return timer === 0 ? 'Lasagna is done.' : 'Not done, please wait.';
}

export function preparationTime(layers, minutePerLayer) {
  return layers.length * (minutePerLayer ?? 2);
}

export function quantities(layers) {
  let needed = {noodles: 0, sauce: 0};
  layers.forEach(layer => {
    if (layer === 'noodles') {
      needed.noodles += 50;
    }
    if (layer === 'sauce') {
      needed.sauce += 0.2;
    }
  })
  return needed;
}

export function addSecretIngredient(friends, mine) {
  mine.push(friends[friends.length - 1]);
}

export function scaleRecipe(recipe, scale) {
  scale /= 2;
  const scaled = {...recipe};
  for (let ingredient in scaled) {
    scaled[ingredient] *= scale;
  }
  return scaled;
}

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

func PreparationTime(layers []string, minutePerLayers int) int {
	if minutePerLayers == 0 {
		minutePerLayers = 2
	}
	return minutePerLayers * len(layers)
}

func Quantities(layers []string) (grams int, liters float64) {
	for _, v := range layers {
		if v == "sauce" {
			liters += 0.2
		}
		if v == "noodles" {
			grams += 50
		}
	}
	return grams, liters
}

func AddSecretIngredient(friendRecipe, myRecipe []string) {
	myRecipe[len(myRecipe)-1] = friendRecipe[len(friendRecipe)-1]
}

func ScaleRecipe(quantities []float64, portions int) (scaled []float64) {
	scale := float64(portions) / 2.0
	for _, v := range quantities {
		scaled = append(scaled, v*scale)
	}
	return scaled
}
